### DISCUSSION TITLE
SvelteKit + Google Auth: invalid request: both auth code and code verifier should be non-empty

### DISCUSSION THREAD

**Discussion by simonh1000** [supabase/supabase#21183] [Questions]
I created a new repo this morning that only has the code suggestions from https://supabase.com/docs/guides/auth/server-side/creating-a-client?framework=sveltekit and **using a local DB with Google oAuth**. Logging in seemed to work 

I then switched back to a repo I had been working that did not work, and ended up copying across all of the src code from new repo. It continued not to work.

I then went back to this new repo, only to discover that it was now broken! By broken I mean that when the auth callback tries to call 
`supabase.auth.exchangeCodeForSession(code)`
I get an error `AuthApiError: invalid request: both auth code and code verifier should be non-empty`

I'm confident the error is _not_ in the code, because it did work earlier today, but in some part of the broader state. I've tried removing .svelte-kit, `supabase db reset`, even `supabase init` but without success.

One thing I do note is that the anon_key has not changed even despite all my resetting.

Any ideas?

**Comment by CalvinWilkinson** [comment]
I am currently having this issue right now and have been struggling for 2-3 days.  I on the other hand am using deno fresh and I am using the SSR client. No matter what I try, I cannot get it to work.  I have read docs, watched YouTube videos, checked discord server, just tried random things, used ChatGPT and more.

I did come across one thing that mentioned that if running locally, which is where I am having issues, this can be an issue.  But it is very informative and not proven.

**Comment by ChristianMP** [comment]
Google led me here while encountering `Auth: invalid request: both auth code and code verifier should be non-empty` while trying to use a password reset link. For others that may be here on the same quest, I suggest reading [this documentation](https://supabase.com/docs/guides/auth/auth-email-templates#redirecting-the-user-to-a-server-side-endpoint), as that got me solved.

**Comment by wiwa** [comment]
I believe the issue is the incorrect `cookies: CookieMethods` setting, specifically in:
```ts
... = createServerClient(PUBLIC_SUPABASE_URL, PUBLIC_SUPABASE_ANON_KEY, {
    cookies: {
      getAll() {
        return event.cookies.getAll()
      },
      setAll(cookiesToSet) {
        // ...
        cookiesToSet.forEach(({ name, value, options }) =>
          event.cookies.set(name, value, { ...options, path: '/' })
        )
      },
    },
  })
```  

Instead of `getAll` and `setAll`, it should be `get` and `set`. Here's what I changed it to:
```ts
  cookies: {
      get: (name) => {
        return event.cookies.get(name);
      },
      set: (name, value, options) => {
        return event.cookies.set(name, value, { ...options, path: "/" });
      }
    }
```

The error seems to be due to `sb-localhost-auth-token` not being set correctly without the correct `.set()` method, which seems to be called by `supabase.auth.signInWithOAuth`.

**Reply by j4w8n** [reply]
The getAll and setAll methods should only be used with the ssr package, v0.4.0 and up.

**Reply by wiwa** [reply]
Thanks, I was somehow still on v0.3.0

**Comment by EpsilonOrionis** [comment]
I've found a solution. You have to redirect using the headers used to create the supabase client.

**Reply by therealpurplemana** [reply]
Can you post more info about this?

**Reply by EpsilonOrionis** [reply]
When you create the supabase client using the function createClientServer, you pass the headers object, as shown here:

```
const headers = new Headers();
  
const supabase = createServerClient(process.env.SUPABASE_URL!, process.env.SUPABASE_ANON_KEY!, {
  cookies: {
    getAll() {
      return parseCookieHeader(request.headers.get('Cookie') ?? '')
    },
    setAll(cookiesToSet) {
      cookiesToSet.forEach(({ name, value, options }) =>
        headers.append('Set-Cookie', serializeCookieHeader(name, value, options))
      )
    },
  },
});
```
When you get the redirect URL from the signInWithOAuth function call:

```
const { data, error } = await supabase.auth.signInWithOAuth({
  provider,
  options: {
    redirectTo: `http://example.com/auth/callback`,
  },
})
```
You have to redirect using the URL obtained AND the headers object that you used to create the supabase client used to perform the request because supabase handles the code verifier generation by itself so if you redirect without the headers you are just providing the auth code without the verification code.

So the redirect is this:

```
if (data?.url) {
    return redirect(data.url, {headers}); // Reindirizza l'utente
} 
```

**Reply by EpsilonOrionis** [reply]
You just need to adapt the code to the framework you’re using. This code is for Remix, but the error is caused for the same reason.

**Reply by stithwilliam** [reply]
This worked for me, thanks! Also using Remix

**Reply by andupp** [reply]
Also remix user and this solved my issue! Thanks a lot!!

**Reply by dcDalin** [reply]
Thanks, scratching my head

**Reply by IGUNUBLUE** [reply]
Thank you so much! it's works for me. The AI just push a lot code and I was thinking, why!!? I decide to read the doc and search the correct solution.

**Comment by logicminds** [comment]
I think I fixed it in nextjs 14.  Basically, I made two calls to `cookies().getAll()` to make this work.  I probably don't need both.  This seems like a bug in nextjs 14.

```typescript
import { createRouteHandlerClient } from '@supabase/auth-helpers-nextjs'
import { cookies } from 'next/headers'
import { NextResponse } from 'next/server'

import type { NextRequest } from 'next/server'
import type { Database } from '@/types.gen'

export async function GET(request: NextRequest) {
  const requestUrl = new URL(request.url)
  const code = requestUrl.searchParams.get('code')

  if (code) {

    const supabase = createRouteHandlerClient<Database>({ cookies })
    cookies().getAll()

    // For some reason I need to get the cookies here to make the exchangeCodeForSession work
    // Don't know if I also need to get them after but will do anyways
    const response  = await supabase.auth.exchangeCodeForSession(code)
    cookies().getAll()
    // RangeError [ERR_HTTP_INVALID_STATUS_CODE]: Invalid status code: 0
    
    if (response.error) {
      console.error('Error logging in:', response.error.message)
      return NextResponse.json({ error: response.error.message }, { status: 400 });
    }
  }

  // URL to redirect to after sign in process completes
  return NextResponse.redirect(requestUrl.origin)
}
```

AI says:

## Possible Explanation: Next.js Cookies Lazy Evaluation
	•	In Next.js 14 with next/headers, cookies() is a lazy getter.
	•	This means that cookies are not actually read from the request until they are accessed explicitly.
	•	When you call cookies().getAll() before exchangeCodeForSession(code), it forces Next.js to read the cookies before Supabase tries to exchange the code.

## Why Does This Fix the Issue?
1.	Before calling exchangeCodeForSession
	      *	The sb-localhost-auth-token-code-verifier cookie exists but is not yet accessed.
	      *	Supabase expects this cookie when exchanging the code.
	
2.	Calling cookies().getAll() forces Next.js to read cookies from the request.
	      *	This ensures that when exchangeCodeForSession(code) runs, the code_verifier cookie is available in memory.
	      *	Without this, Supabase might think the cookie is missing (because Next.js hasn’t read it yet).
	
3.	After the exchange attempt, cookies are read again.
	      *	Supabase successfully clears the sb-localhost-auth-token-code-verifier cookie after the exchange.

**Comment by magick93** [comment]
I'm getting the `AuthApiError: invalid request: both auth code and code verifier should be non-empty` error, but only in Firefox. 

I'm using Sveltekit.

**Comment by SqlOwl** [comment]
I am getting this issue in Next.JS 15.1.7 application on Google Auth Callback when calling supabase.auth.exchangeCodeForSession(code):

`Exchange code error: invalid request: both auth code and code verifier should be non-empty`

This problem is intermittent, sometimes it is ok and sometimes it is there for some users. I have went through solutions on this page and nothing seems to stop the issue from happening time to time. I have Verified google Authentication for production and it is there that it happens the most, really enjoying. What I have figured out is that when it fails supabase auth token code verifier is missing. Can anyone help please! I am following standard code from Supabase for middleware, client and callback

**Reply by SqlOwl** [reply]
I have seem to solve my issue. It was mistake on my side of not accounting for https://**WWW**.mydomain.com For Both Supabase and Google this is a separate domain and any redirect to it is blocked. Although not blocked but no code verifier is returned

**Reply by chrisdemetriad** [reply]
What, in the Supabase Redirect URL, or..

**Reply by therealpurplemana** [reply]
> What, in the Supabase Redirect URL, or.. 
> 
> 

It will depend. That won't solve it for many of the cases described here. Describe your setup in detail to help us help debug.

**Reply by keremnalbant** [reply]
Same, it happens for some users / devices. Most of the time it works but some times it doesn't return the code-verifier. I can confirm that I have the correct setup in terms of redirect uris and www case. I'm on Next.js 15. Please let me know if you have any clue or how to debug this or get a more explanatory error message from Google's or Supabase's side.

**Comment by qiweiii** [comment]
My root cause was different client and server client auth options, after migrate to the new `@supabase/ssr` package, we have to rmb to use the same auth options:

```ts
createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      auth: commonAuthOptions, // use a common one
```

**Comment by wu0407** [comment]
I have the same issue in local development with local deploy supabase. The root cause is that my local site is HTTP://127.0.0.1:3000, and the OAuth callback URL is http://xxxx.xxxx.  The xx-code-verifier cookie is not on the callback URL.

The solution is to use the outside callback URL that can be accessed. So the xx-code-verifier cookie will exist.

**Comment by Oussail** [comment]
Hello everybody ✋

✅ **The solution to the issue! Save yourself days of headache** 🧠💥

> 🛠️ This solution is for **anyone experiencing** the infamous error below specifically when trying to **reset a password** in **Next.js 14/15+**:

```
invalid request: both auth code and code verifier should be non-empty
```

If you're dealing with this error, then *forget about chasing down `xx-code-verifier`*. It's most likely **getting lost somewhere between Supabase and your middleware/callback setup**. After hours (read: *days*) of debugging, I finally found a **reliable solution** that works 🎉

---

## 💡 Solution: Use Supabase's `verifyOtp` flow instead

### 1. Trigger password reset with `resetPasswordForEmail`

To send to the user receives the reset link, trigger this function in your frontend:

```ts
await supabase.auth.resetPasswordForEmail(email);
```

---

### 2. Create a route `/auth/confirm/route.ts` to handle the redirect + verification:

```ts
import { type EmailOtpType } from "@supabase/supabase-js";
import { NextRequest, NextResponse } from "next/server";
import { createClient } from "@/utils/supabase/server";

export async function GET(request: NextRequest) {
  const { searchParams } = new URL(request.url);
  const token_hash = searchParams.get("token_hash");
  const type = searchParams.get("type") as EmailOtpType | null;
  const next = searchParams.get("next") ?? "/";
  const redirectTo = request.nextUrl.clone();
  redirectTo.pathname = next;

  if (token_hash && type) {
    const supabase = await createClient();
    const { error } = await supabase.auth.verifyOtp({ type, token_hash });

    if (!error) {
      return NextResponse.redirect(redirectTo);
    }

    console.log("🚀 ~ GET ~ error:", error);
  }

  // Redirect to error page if something went wrong
  redirectTo.pathname = "/auth/auth-code-error";
  return NextResponse.redirect(redirectTo);
}
```

---

### 3. Update your Supabase password reset email template (Supabase Dashboard → Auth → Templates → Reset Password):

You can find it at:  
`https://supabase.com/dashboard/project/[YOUR_PROJECT_ID]/auth/templates`

Update the reset password template with the following HTML:

```html
<h2>Reset Password</h2>
<p>Follow this link to reset your password:</p>
<p>
  <a href="{{ .SiteURL }}/auth/confirm?token_hash={{ .TokenHash }}&type=recovery&next=/account/update-password">
    Reset Password
  </a>
</p>
```

---

### 4. Create the page `/account/update-password/page.tsx` where users will land after verification to actually update their password.

This is where you prompt the user to enter a new password.

---

📚 **Docs for reference**:  
[https://supabase.com/docs/guides/auth/passwords?queryGroups=flow&flow=pkce#resetting-a-password](https://supabase.com/docs/guides/auth/passwords?queryGroups=flow&flow=pkce#resetting-a-password)

---

Hope this helps someone.  
Let me know if you have questions!

**Reply by sarperdag** [reply]
I bet you wrote this using ChatGPT or something. It doesn't work.

**Reply by mt816** [reply]
You’re a lifesaver! This fix worked great on my end too. Thanks a ton for sharing 🙌

**Comment by macs1756** [comment]
Hey, may be heplfull, what i did i this sitution
1) Added in config with creating client this:
```
{
      auth: {
        flowType: "pkce",
      },
    }
```
 2) Generate mail:
 ``` 
await supabase.auth.resetPasswordForEmail(
      email,
      {
        redirectTo: "https//... this url need to be set up on supabase dashboard",
      }
    );
```
  3) In page on server rendering generate session and update user(code below it's example you can modify it)
  ```
const cookieStore = await cookies();

  const base64Token = cookieStore.get(
    "sb-iangjclvbrgqrgzonrob-auth-token"
  ).value;

  const jsonPart = base64Token.replace("base64-", "");
  const decoded = atob(jsonPart);
  const sessionObj = JSON.parse(decoded);

  const supabase = await createClient();

  const { data, error } = await supabase.auth.setSession({
    access_token: sessionObj.access_token,
    refresh_token: sessionObj.refresh_token,
  });

  const { error: error2 } = await supabase.auth.updateUser({
    password: "000000",
  });
```
  You may have other cookies and you can use their

**Comment by EskelCz** [comment]
This happens when one browser is used for sending the email and another when opening the confirmation link. Why is a verifier is needed, when for the otp code version it's not?
